src/test/sync_tests.cpp

```diff
@@ -29,6 +32,38 @@ void TestPotentialDeadLockDetected(MutexType& mutex1, MutexType& mutex2)
     BOOST_CHECK(!error_thrown);
     #endif
 }
+
+#ifdef DEBUG_LOCKORDER
+template <typename MutexType>
+void TestDoubleLock2(MutexType& m)
+{
+    ENTER_CRITICAL_SECTION(m);
+    LEAVE_CRITICAL_SECTION(m);
+}
+
+template <typename MutexType>
+void TestDoubleLock(bool should_throw)
+{
+    const bool prev = g_debug_lockorder_abort;
+    g_debug_lockorder_abort = false;
+
+    MutexType m;
+    ENTER_CRITICAL_SECTION(m);
+    if (should_throw) {
+        BOOST_CHECK_EXCEPTION(
```

Comment:
Could limit the `#ifdef DEBUG_LOCKORDER` to just this one line/statement? Benefit would be that `BOOST_CHECK_NO_THROW` will be run for non-debug builds as well.

Reply:
Without `DEBUG_LOCKORDER` we would actually do the double lock and once executed then we are in "undefined behavior" state. We don't want to do that because it may randomly crash the test executable, overwrite random portions of memory for the excitement of subsequent tests, etc.

https://github.com/bitcoin/bitcoin/blob/b1f94072a1cc3ca67ce13658bb981d03480ef38e/src/test/sync_tests.cpp#L93-L95

Reply:
I compiled and tested locally with sanitizers and it passed. Am I missing something? https://github.com/bitcoin/bitcoin/pull/19337#issuecomment-669211924